From 464c5b5df578b0ea7b9a9a658926c8a0adebaaa8 Mon Sep 17 00:00:00 2001
From: vrp <vijaykumar.rp@intel.com>
Date: Wed, 23 Feb 2022 10:11:47 +0100
Subject: Add extmark match based classification support in tc filter

Enables "tc filter" QoS Classify operations based on extmark value.

---
 include/uapi/linux/pkt_cls.h |  1 +
 tc/f_u32.c                   | 49 ++++++++++++++++++++++++++++++++++++
 2 files changed, 50 insertions(+)

diff --git a/include/uapi/linux/pkt_cls.h b/include/uapi/linux/pkt_cls.h
index 6836ccb9..28938439 100644
--- a/include/uapi/linux/pkt_cls.h
+++ b/include/uapi/linux/pkt_cls.h
@@ -230,6 +230,7 @@ enum {
 	TCA_U32_MARK,
 	TCA_U32_FLAGS,
 	TCA_U32_PAD,
+	TCA_U32_EXTMARK,
 	__TCA_U32_MAX
 };
 
diff --git a/tc/f_u32.c b/tc/f_u32.c
index a5747f67..227b52a1 100644
--- a/tc/f_u32.c
+++ b/tc/f_u32.c
@@ -690,6 +690,41 @@ static int parse_mark(int *argc_p, char ***argv_p, struct nlmsghdr *n)
 	return res;
 }
 
+static int parse_extmark(int *argc_p, char ***argv_p, struct nlmsghdr *n)
+{
+       int res = -1;
+       int argc = *argc_p;
+       char **argv = *argv_p;
+       struct tc_u32_mark extmark;
+
+       if (argc <= 1)
+               return -1;
+
+       if (get_u32(&extmark.val, *argv, 0)) {
+               fprintf(stderr, "Illegal \"extmark\" value\n");
+               return -1;
+       }
+       NEXT_ARG();
+
+       if (get_u32(&extmark.mask, *argv, 0)) {
+               fprintf(stderr, "Illegal \"extmark\" mask\n");
+               return -1;
+       }
+       NEXT_ARG();
+
+       if ((extmark.val & extmark.mask) != extmark.val) {
+               fprintf(stderr, "Illegal \"extmark\" (impossible combination)\n");
+               return -1;
+       }
+
+       addattr_l(n, MAX_MSG, TCA_U32_EXTMARK, &extmark, sizeof(extmark));
+       res = 0;
+
+       *argc_p = argc;
+       *argv_p = argv;
+       return res;
+}
+
 static int parse_selector(int *argc_p, char ***argv_p,
 			  struct tc_u32_sel *sel, struct nlmsghdr *n)
 {
@@ -727,6 +762,9 @@ static int parse_selector(int *argc_p, char ***argv_p,
 	} else if (matches(*argv, "mark") == 0) {
 		NEXT_ARG();
 		res = parse_mark(&argc, &argv, n);
+	} else if (matches(*argv, "extmark") == 0) {
+		NEXT_ARG();
+		res = parse_extmark(&argc, &argv, n);
 	} else if (matches(*argv, "ether") == 0) {
 		NEXT_ARG();
 		res = parse_ether(&argc, &argv, sel);
@@ -1291,6 +1329,17 @@ static int u32_print_opt(struct filter_util *qu, FILE *f, struct rtattr *opt,
 		}
 	}
 
+	if (tb[TCA_U32_EXTMARK]) {
+		struct tc_u32_mark *mark = RTA_DATA(tb[TCA_U32_EXTMARK]);
+
+		if (RTA_PAYLOAD(tb[TCA_U32_EXTMARK]) < sizeof(*mark)) {
+			fprintf(f, "\n  Invalid extmark (kernel&iproute2 mismatch)\n");
+		} else {
+			fprintf(f, "\n  extmark 0x%04x 0x%04x (success %d)",
+				mark->val, mark->mask, mark->success);
+		}
+	}
+
 	if (sel) {
 		if (sel->nkeys) {
 			int i;
